### DISCUSSION TITLE
Does pattern changed about using selector in component?

### DISCUSSION THREAD

**Discussion by seanbruce** [pmndrs/zustand#971] [General] [answered]
Hi, I do remember the document used to describe how to use selectors in components to improve performance:
```js 
// selectors don't rely on states or props

const selector = (state) => state.name

function App() {
    const name = useStore(selector);
    // ...
}
```

or

```js
// selectors do rely on states or props

function App({id}) {
  const selector = useCallback((state) => {
    return state[id]
  }, [id]);
  const state = useStore(selector);
  // ...
}
```

now those cannot be found in the document.
My question is whether the selector no longer needs to be stable to improve performance. because I used to follow those patterns in my code strictly, if the function stability doesn't matter anymore, I could save a few keystrokes.

**Comment by dai-shi** [comment, marked as answer]
Thanks for opening a discussion.

In the past, we recommended the use of stable selectors in v3, because we were targeting v4 with useMutableSource at that point.

After that, React deprecated experimental useMutableSource and provides useSyncExternalStore. 
For more context:
1. https://twitter.com/TkDodo/status/1427922647401312257
2. https://github.com/reactwg/react-18/discussions/84
3. https://github.com/reactwg/react-18/discussions/86

Now, zustand v4 is based on useSyncExternalStore, so stable selectors are not necessary. But if you have a stable selector, you can expect the performance is a little more improved.

In summary:

## inline selectors are just fine and recommended
```js
const Component = () => {
  const slice = useStore((state) => state.slice);
  // ...
```

## useCallback for lightweight selectors is an extra overhead, and not recommended
```js
const Component = () => {
  const slice = useStore(useCallback((state) => state.slice, [])); // ❌ 
  // ...
```

## Having stable selectors can improve performance, but not very noticeable

```js
const selector = (state) => state.slice; // ✅ 
const Component = () => {
  const slice = useStore(selector);
  // ...
```

**Reply by seanbruce** [reply]
Thanks for your summary. It cleared my confusion. I will prefer the inline selector now.

**Reply by pastelmind** [reply]
@dai-shi  I read the Zustand docs and this seems to go against the [recipe here](https://github.com/pmndrs/zustand/blob/1dd85fe24d35d28b966840a520fef1b2082a7ba3/docs/recipes/recipes.mdx#memoizing-selectors):

> It is generally recommended to memoize selectors with `useCallback`. This will prevent unnecessary computations each render. It also allows React to optimize performance in concurrent mode.

Is this advice no longer accurate?

More generally, does Zustand skip computing the new value across re-renders if the store is unchanged AND the identity of the selector function is the same? Knowing this would provide better insight into whether I should use `useCallback()` for a particular selector.

**Reply by dai-shi** [reply]
That's no longer accurate. The recipes.mdx is a copy of readme.md at once, and not well maintained. Can you open a PR to fix?

**Reply by pastelmind** [reply]
> That's no longer accurate. The recipes.mdx is a copy of readme.md at once, and not well maintained. Can you open a PR to fix?

Sure.

I edited a new question into my previous reply, which you may have missed b/c you replied so quickly. Could you please check it as well?

**Reply by dai-shi** [reply]
A related PR https://github.com/pmndrs/zustand/pull/550/files#diff-5a831ea67cf5cf8703b0de46901ab25bd191f56b320053be9332d9a3b0d01d15. We can simply remove the section.

**Reply by pastelmind** [reply]
Ah, it appears that Zustand delegates selector management to `useSyncExternalStoreWithSelector()`:

https://github.com/pmndrs/zustand/blob/1dd85fe24d35d28b966840a520fef1b2082a7ba3/src/react.ts#L35-L49

I checked the [implementation of `useSyncExternalStoreWithSelector()`](https://github.com/facebook/react/blob/ffb6733eefc23e62c1fb94ce7fc442f87255ebd7/packages/use-sync-external-store/src/useSyncExternalStoreWithSelector.js) from the `use-sync-external-store` package. This hook uses `useMemo()` to compute the next slice, and since the selector is in the dependency array, it will re-run when the selector identity changes. However, a well-behaved selector (or one with an appropriate equality function) would allow React to bail out from re-rendering anyway. The overall cost of invalidating `useMemo()` would be negligible in light of everything else while rendering.

I also read https://github.com/reactwg/react-18/discussions/86 which lead me to https://github.com/reactwg/react-18/discussions/84, which gives me hope that the React team is committed to ensuring that selector stability would not become a performance pitfall in the future.

Edit: And the answer to my own question

> More generally, does Zustand skip computing the new value across re-renders if the store is unchanged AND the identity of the selector function is the same? Knowing this would provide better insight into whether I should use `useCallback()` for a particular selector.

appears to be false. Both the genuine and shimmed versions of `useSyncExternalStore()` calls the selector on every render, regardless of whether it is stable:

https://github.com/facebook/react/blob/ffb6733eefc23e62c1fb94ce7fc442f87255ebd7/packages/react-reconciler/src/ReactFiberHooks.js#L1679-L1682

https://github.com/facebook/react/blob/ffb6733eefc23e62c1fb94ce7fc442f87255ebd7/packages/use-sync-external-store/src/useSyncExternalStoreShimClient.js#L53-L57

So a selector is called on every render regardless of whether it is stable (or memoized). I assume the general advice would be:

1. Don't bother with memoizing selectors in most cases
2. Move expensive computations out of selectors and into your own `useMemo()`s

**Reply by pastelmind** [reply]
Relevant discussion here: https://github.com/pmndrs/zustand/discussions/658

**Reply by dai-shi** [reply]
Wow, you investigate it pretty well.

> Don't bother with memoizing selectors in most cases

Correct.

> Move expensive computations out of selectors and into your own useMemo()s

For single component use case, it's true.
For shared selector across component use case, memoization with `reselect`, `proxy-memoize` or such would help.

**Reply by LiangMingChen** [reply]
the latest version now the getSnapShot is always new function per render.
https://github.com/pmndrs/zustand/blob/41fed0fe6f190472bb6b331584cf087546f9df33/src/react.ts#L26
<img width="558" height="252" alt="image" src="https://github.com/user-attachments/assets/33448065-5bfc-4af3-8f14-4c61a7509022" />


This now is always get execute inside the useSyncExternalStore
Yes it does not cause another `force` render, 
<img width="667" height="325" alt="image" src="https://github.com/user-attachments/assets/66898df3-3bfa-4cee-b4e0-84e20475954c" />

useSyncExternalStoreWithSelector also rely on getSnapshot stable to provide the stable getSelection
Both value and getSnapShot trigger useLayoutEffect.  I understand value should be stable and lightweight so calling it should not be big manner.  The fact that it is not duce to subscribe notification change that would trigger extra work useLayoutEffect or Fiber equvalient.
https://github.com/facebook/react/blob/19baee813cb08fc5971a3b301814faa6cd530ec8/packages/react-reconciler/src/ReactFiberHooks.js#L1705C1-L1716C5
https://github.com/facebook/react/blob/19baee813cb08fc5971a3b301814faa6cd530ec8/packages/react-reconciler/src/ReactFiberHooks.js#L1833C1-L1853C2


I might miss something, -- Is there a reason not keep getSnapshot react stable?

**Reply by dbritto-dev** [reply]
@LiangMingChen what needs to be stable is the snapshot result not the snapshot function itself

**Reply by dai-shi** [reply]
I think it's a nice catch. #3192

It increases a bundle size a little bit though.

**Reply by LiangMingChen** [reply]
@dai-shi   Thanks for fixing this
